api/src/rate_limit.py

```diff
@@ -0,0 +1,242 @@
+        :param user_id: A string uniquely identifying the user, should be the
+        same across servers like the IP address of the user
+        :raises TooManyConnectionsException if the user already has
+        MAX_CONNECTIONS_PER_USER active connections
+        """
+        ...
+
+    async def release_connection(self, user_id: str) -> None:
+        """
+        Release a connection for the given user id
+        :param user_id: A string uniquely identifying the user, should be the
+        same across servers like the IP address of the user
+        """
+        ...
+
+    def rate_limited_connection(self, user_id: str) -> AsyncContextManager:
+        """
+        Reserve a connection for the given user_id for the duration of the
+        context
+        :param user_id: A string uniquely identifying the user, should be the
+        same across servers like the IP address of the user
+        :raises TooManyConnectionsException if the user already has
+        MAX_CONNECTIONS_PER_USER active connections
+        """
+        ...
+
+    async def refresh_server_liveness(self, user_ids: Iterator[str]) -> None:
+        """
+        This function should be called every
+        SERVER_LIVENESS_EXPIRATION_SECONDS/3 while the server is operating
+        """
+        ...
+
+
+_USER_LIVENESS_REFRESH_BATCH_SIZE = 50
+
+# Lua script to increment the connection count for a given user on the given
+# server, if the user has not exceeded their connection count across all
+# servers.
+# Returns true if user has not exceeded their max connection count and the
+# new connection has been recorded, false otherwise
+# language=lua
+ACQUIRE_CONNECTION_SLOT = f"""
+local user_id = ARGV[1]
+local target_server_id = ARGV[2]
+
+local user_connections_key = 'user-connections:' .. user_id
+local server_ids = redis.call('hkeys', user_connections_key)
+
+local total_connection_count = 0
+for _, server_id in pairs(server_ids) do
+    local server_expired = not redis.call('get', 'api-server:' .. server_id)
+    if server_expired then
+        redis.call('hdel', user_connections_key, server_id)
+    else
+        local server_connection_count = redis.call(
+            'hget',
+            user_connections_key,
+            server_id
+        )
+        if server_connection_count ~= false then
+            total_connection_count = total_connection_count + server_connection_count
+        end
+    end
+end
+
+if total_connection_count < {MAX_CONNECTIONS_PER_USER} then
+    redis.call('hincrby', user_connections_key, target_server_id, 1)
+    redis.call('expire', user_connections_key, {SERVER_LIVENESS_EXPIRATION_SECONDS})
+    return true
+else
+    return false
+end
+"""
+
+# language=lua
+RELEASE_CONNECTION_SLOT = """
+    local user_id = ARGV[1]
+    local server_id = ARGV[2]
+
+    local user_key = 'user-connections:' .. user_id
+    local count = tonumber(redis.call('hget', user_key, server_id))
+    if count ~= nil and count > 0 then
```

Comment:
Not sure in what way you're asking.

This is decrementing the connection count and it exists, line 83 is adding the connection to the total count if it exists.

If you're asking why I compare to false on line 83 and nil here, you're right, I should be comparing to nil on line 83

Reply:
Yeah, comparing to false vs this was what I was asking about.

Reply:
I noticed I'm using tonumber here but not on line 83. It worked in line 83 because `2 + "3" == 5` in lua. But didn't work here because `2 < "3"`  results in a type error. I changed 83 to use tonumber for consistency.